precompile/testutils/test_precompile.go

```diff
@@ -69,18 +98,76 @@ func (test PrecompileTest) Run(t *testing.T, module modules.Module, state contra
 		input = test.InputFn(t)
 	}
 
-	if input != nil {
-		ret, remainingGas, err := module.Contract.Run(accesibleState, test.Caller, contractAddress, input, test.SuppliedGas, test.ReadOnly)
-		if len(test.ExpectedErr) != 0 {
-			require.ErrorContains(t, err, test.ExpectedErr)
-		} else {
-			require.NoError(t, err)
-		}
-		require.Equal(t, uint64(0), remainingGas)
-		require.Equal(t, test.ExpectedRes, ret)
+	return PrecompileRunparams{
+		AccessibleState: accesibleState,
+		Caller:          test.Caller,
+		ContractAddress: contractAddress,
+		Input:           input,
+		SuppliedGas:     test.SuppliedGas,
+		ReadOnly:        test.ReadOnly,
+	}
+}
+
+func (test PrecompileTest) Bench(b *testing.B, module modules.Module, state contract.StateDB) {
+	runParams := test.setup(b, module, state)
+
+	if runParams.Input == nil {
+		b.Skip("Skipping precompile benchmark due to nil input (used for configuration tests)")
 	}
 
+	stateDB := runParams.AccessibleState.GetStateDB()
+	snapshot := stateDB.Snapshot()
+
+	ret, remainingGas, err := module.Contract.Run(runParams.AccessibleState, runParams.Caller, runParams.ContractAddress, runParams.Input, runParams.SuppliedGas, runParams.ReadOnly)
+	if len(test.ExpectedErr) != 0 {
+		require.ErrorContains(b, err, test.ExpectedErr)
+	} else {
+		require.NoError(b, err)
+	}
+	require.Equal(b, uint64(0), remainingGas)
+	require.Equal(b, test.ExpectedRes, ret)
+
 	if test.AfterHook != nil {
-		test.AfterHook(t, state)
+		test.AfterHook(b, state)
+	}
+
+	b.ReportAllocs()
+	start := time.Now()
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		// Revert to the previous snapshot and take a new snapshot, so we can reset the state after execution
+		stateDB.RevertToSnapshot(snapshot)
+		snapshot = stateDB.Snapshot()
+
+		// Ignore return values for benchmark
+		_, _, _ = module.Contract.Run(runParams.AccessibleState, runParams.Caller, runParams.ContractAddress, runParams.Input, runParams.SuppliedGas, runParams.ReadOnly)
+	}
+	b.StopTimer()
```

Comment:
at here, aren't we also including the time for reverting to the snapshot?

Reply:
The original iteration of this used `StartTimer` and `StopTimer`, but this resulted in an interesting bug.

The golang benchmarks use the execution time tracked during an iteration of the for loop in order to figure out how many times it should re-run the tests to get an accurate benchmark.

This means that if the time spent in the for loop is extremely small, it will end up running the loop a much larger number of times.

A strange result of this pattern is that benchmarks need to be sure that a majority of the time spent in each iteration of the loop is tracked. Otherwise, you can run into an issue where a single iteration takes 1ns, so the benchmark decides to perform it 1 million times. However, if the setup/teardown in the for loop actually takes 1 second, then it will still execute the benchmark 1 million times and it will execute the setup/teardown code a large number of times as well.

The original iteration of this called `StartTimer` and `StopTimer`, but it turned out that capturing the memory statistics took a much longer amount of time than the execution that was being benchmarked, so the benchmarks actually took minutes to complete.

Here's a more intuitive example snippet:
```go
func BenchmarkSleepBenchmark(b *testing.B) {
	for i := 0; i < b.N; i++ {
		time.Sleep(time.Nanosecond)
		// time.Sleep(time.Millisecond)
		b.StopTimer()
		b.StartTimer()
	}
}
```

Reply:
hmm yea weird behaviour, it's probably expected to use them before/after running each iteration (not inside the loop).

Reply:
These are actually used fairly regularly inside of the for loop in benchmarks, they just run into issues if the computation being benchmarked is extremely small.